### ISSUE TITLE
No Devices found on Android 6 or higher

### ISSUE THREAD

**Issue by DumDumin** [xabre/xamarin-bluetooth-le#104]
On Android 6 and 7 there is no device found. Android 4 and 5 are working fine.
Does anyone else has this problem?

**Comment by xabre** [comment]
@DumDumin 

From our readme : )
Add these permissions to AndroidManifest.xml. For Marshmallow, please follow [Requesting Runtime Permissions in Android Marshmallow](https://blog.xamarin.com/requesting-runtime-permissions-in-android-marshmallow/) 

```xml
<uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
<uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
<uses-permission android:name="android.permission.BLUETOOTH" />
<uses-permission android:name="android.permission.BLUETOOTH_ADMIN" />
```

Android >=6 requires location access and needs the user to grant this permission. Follow the link to see how this is done. In the future this might become part of the plugin, but for now this is up to the developer.

**Comment by AmperayaniMadan** [comment]
I too am facing a similar issue. Does not discover the BLE device in Android 6.0. I have not tested in earlier versions.

I also tried the above mentioned permissions, does not work even then.

One thing i noticed is, outside of my application, when I turn on bluetooth I am able to see my BLE device in the result list. But why does it not work in my app?

**Comment by DumDumin** [comment]
For me its only working if GPS is turned on, is that right?

**Comment by AmperayaniMadan** [comment]
Glad it is working for you after turning on the GPS.

Its not working for me, even when the GPS is turned on nothing happens, the OnDeviceDiscovered() is not being hit. I am searching for a specific device using it UUID.

Im running on Samsung Galaxy Tab S2 with OS 6.0.1

**Comment by xabre** [comment]
@AmperayaniMadan If you don't want to implement the permission requests like described in [Requesting Runtime Permissions in Android Marshmallow](https://blog.xamarin.com/requesting-runtime-permissions-in-android-marshmallow/),  you have to manually turn it on for your app form Settings>Apps>YourApp>Permissions. I've attached a screenshot from our ble sample app:

![permissions](https://cloud.githubusercontent.com/assets/6204446/19560367/807b5380-96d4-11e6-9a01-954634cb135e.png)

@DumDumin That is correct, Android >=6 need location to be enabled for ble scanning. This is because of beacons. :P You can read more about this on the android official doc.

**Comment by AmperayaniMadan** [comment]
@xabre 
I did check the location and BT permissions in the App Manager for my App, they are enabled.

But what is curious is, turning on the BT manually, outside of my App, i am able to see my BLE device in the list, at this instant the GPS is not turned on. I tripple checked this.

:| :|

**Comment by xabre** [comment]
@AmperayaniMadan are you sure it's a bluetooth low energy device (GATT) device? Maybe you can install the 'BLE Scanner' app from the play store and see if it can detect your device?

**Comment by AmperayaniMadan** [comment]
@xabre
Yes it is a GATT device, I am able to see it in the BLE Scanner app and also connect and see its characteristics.

My code works fine for iOS, but when it comes to Android it flops. May be I am missing something apart from the permissions. I will share my code, if anyone is kind enough to point out what I am missing that will be very helpful.

But, can you share any sample code which works for Android 6.0.1. I can quickly try that and rectify.

Thanks

**Comment by xabre** [comment]
Well the sample app works fine on my Android 6 phone together with the screenshot above :).
Just post your code snippet here :) and I'll try to help.

**Comment by AmperayaniMadan** [comment]
Attaching a snippet of the method which I use to connect to and get data from my BLE device.

Will provide the output log in sometime.

![ble_codesnip](https://cloud.githubusercontent.com/assets/19404703/19595278/b21f6a3e-97a6-11e6-9474-9c6e76e1f430.jpg)

**Comment by xabre** [comment]
Hmm... you are calling StartScan with a guid as a parameter. You have to make sure that the peripheral includes the service guid in it's advertisement package. 
As a test, try removing the GUID parameter and let the scan run without any parameters.

**Comment by AmperayaniMadan** [comment]
:), I tried that too earlier, didn't work. But my colleagues have a different model of the same Samsung Galaxy Tab S2, and they are able to connect to the BLE device. What I got to know is in their model they do not have a provision to turn off the GPS (weird !!, no privacy eh?).

But that should not be any different than manually turning on the GPS !!

We haven't implemented the runtime permissions in our app (but it should not be of any effect until you revoke the permissions right ?).

**Comment by xabre** [comment]
- What does you application output window say? If the permission is not ok it will be reflected there.
- Did you try the sample app? (you also have to grant it location permissions)

**Comment by PawelGasiorowski** [comment]
I have the same problem. In my case I can connect the device on both Android and IOS systems, hovewer Android does not contain the device name which is null. Galaxy S7.

**Comment by AbanoubSedrakDev** [comment]
Device Name Null while address  has a value

**Comment by bed007** [comment]
If I am not mistaken, the Device Name is populated from the advertisement packets. So is it possible that your BLE device does not provide a name ?

Are you using a commercial device or are you developping your own ?

**Comment by PawelGasiorowski** [comment]
Hi,

Update from my side. As deep as I dug into this, the device name might be passed through ADV record data, however it is byte limited. In a result, most of devices that I tested didn't have it. IOS has built in mechanism to handle this - it has some system cache, not possible to clear, that pairs device IDs (mac address accessible from system layer or simple GUID) with the name that IOS gets automatically from Device info characteristic just after first connection. Result is that even if the ADV record does not contain device name, IOS knows what it is, looking at the cache. Of course if no connection has been established IOS can't use the magic to guess the name, but it is good to know this cache exists when testing any solution :)

Android doesn't have a cache, so the name is always empty.
I ended up coding the cache inside the application to cover Android.

cheers,
Pawel

**Comment by Malik-Jehangir** [comment]
@PawelGasiorowski  How did u do it? Can u brief me cuz Im new to this and my application opens on android, although my bluetooth is on but I still unable to see my polar H7 heart rate monitor?

**Comment by Malik-Jehangir** [comment]
Hi all

Ive got it working, it is showing all the characteristics of the heart rate monitor except the heart rate.. Any idea why?

**Comment by bapgobears** [comment]
> @DumDumin
> 
> From our readme : )
> Add these permissions to AndroidManifest.xml. For Marshmallow, please follow [Requesting Runtime Permissions in Android Marshmallow](https://blog.xamarin.com/requesting-runtime-permissions-in-android-marshmallow/)
> 
> ```
> <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
> <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
> <uses-permission android:name="android.permission.BLUETOOTH" />
> <uses-permission android:name="android.permission.BLUETOOTH_ADMIN" />
> ```
> 
> Android >=6 requires location permission in order to perform BLE discovery (this is because of iBeacons) and needs the user to grant this permission. Just adding those lines to your manifest without permission request **will not be enough.** You still need to go to Apps>YourApp>Permissions and enable Location. Follow the link to see how this is done. In the future this might become part of the plugin, but for now this is up to the developer.

Hi Guys,

I think we could save a lot of headache for people getting started if on the FRONT PAGE of the repository after describing the MANIFEST changes for Android, add the sentence, "Android >=6 REQUIRES you to enable Location permissions on your device for your app.  Go to Settings/Apps/YourApp/Permissions and enable Location".  If you don't do this, the scan will not find any devices.

Thanks!
Brett

**Comment by Kikanye** [comment]
> Hi,
> 
> Update from my side. As deep as I dug into this, the device name might be passed through ADV record data, however it is byte limited. In a result, most of devices that I tested didn't have it. IOS has built in mechanism to handle this - it has some system cache, not possible to clear, that pairs device IDs (mac address accessible from system layer or simple GUID) with the name that IOS gets automatically from Device info characteristic just after first connection. Result is that even if the ADV record does not contain device name, IOS knows what it is, looking at the cache. Of course if no connection has been established IOS can't use the magic to guess the name, but it is good to know this cache exists when testing any solution :)
> 
> Android doesn't have a cache, so the name is always empty.
> I ended up coding the cache inside the application to cover Android.
> 
> cheers,
> Pawel

How did you do this?

**Comment by PawelGasiorowski** [comment]
Use device mac address as a key (available only in android) and name from device Info characteristic as a value. Alternatively you can check if Android doesn't do it automatically from deviceInfo.

**Comment by janusw** [comment]
I think this issue can be closed. The README explains the runtime permissions requirements for Android 6+. For any further problems discussed here, please open a new issue if necessary.

**Comment by janseris** [comment]
Not only you need permission but also turned on GPS to find any device.
